src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,259 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from datetime import datetime, timedelta
+from typing import Any, Generic, Sequence, TypeVar, Union
+
+import numpy as np
+
+T = TypeVar("T")
+
+Container = Union[list, np.ndarray]
+
+
+class OrderedRingBuffer(Generic[T]):
+    """Time aware ringbuffer that keeps its entries sorted by time."""
+
+    def __init__(
+        self,
+        buffer: Any,
+        sample_rate: float,
+        window_border: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """Initialize the time aware ringbuffer.
+
+        Args:
+            buffer: instance of a buffer container to use internally
+            sample_rate: resolution of the incoming timestamps in
+                hz (samples per second)
```

Comment:
maybe I missed something,  but didn't we settle on sampling interval (or period)?

Reply:
@matthias-wende-frequenz argued hard for this, argument was that it is used in the rest of the sdk... I personally would rather have the interval myself...

Reply:
I argued for it until we had a discussion and everyone else favored `resampling_period_s` as used in the resampler.